### ISSUE TITLE
TypeError in constructor of class OrderShippingAddressForViewing on null data in table ps_address

### ISSUE THREAD

**Issue by FredBrockstedt** [PrestaShop/PrestaShop#23336]
####  Observed
BackOffice fails with 500 and order can not be processed.

####  Expected
process order without error

####  How to reproduce

1. Go to the BackOffice > Orders > Orders 
2. click on an order where one of the following columns is NULL in the ps_address: phone, company, address2, phone_mobile
3. See error

**Cause**
NULL is not a string, therefor arguments passed to the OrderShippingAddressForViewing constructor raise TypeError

**Possible solutions**
1. when setting up prestashop force NOT NULL on optional address fields, will only work for new installations
2. make the constructor of OrderShippingAddressForViewing more robust against other datatypes
3. propper database migrations

**hot-fix**
I ran this on the mysql server, for the moment it's working again
```
UPDATE `ps_address` SET `phone`="" WHERE `phone` IS NULL; 
UPDATE `ps_address` SET `company`="" WHERE `company` IS NULL;
UPDATE `ps_address` SET `address2`="" WHERE `address2` IS NULL;
UPDATE `ps_address` SET `phone_mobile`="" WHERE `phone_mobile` IS NULL;

ALTER TABLE `ps_address` MODIFY `phone` VARCHAR(32) NOT NULL;
ALTER TABLE `ps_address` MODIFY `company` VARCHAR(255) NOT NULL;
ALTER TABLE `ps_address` MODIFY `address2` VARCHAR(128) NOT NULL;
ALTER TABLE `ps_address` MODIFY `phone_mobile` VARCHAR(32) NOT NULL;
```

**Full error**
`Type error: Argument 4 passed to PrestaShop\PrestaShop\Core\Domain\Order\QueryResult\OrderShippingAddressForViewing::__construct() must be of the type string, null given, called in /var/www/vhosts/<hostname redacted>/httpdocs/src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php on line 270`

#### Additional infos
PrestaShop version: 1.7.7.1
PHP version: 7.2

**Comment by hibatallahAouadni** [comment]
Hello @FredBrockstedt 

Could you please provide PS version and your PHP version, and did you upgrade your shop lately?
Please check and feedback.

Thanks!

**Comment by FredBrockstedt** [comment]
Hi @hibatallahAouadni,

this was on Prestashop 1.7.7.1 and PHP 7.2 (using nginx and fpm). Yes I updated lately.
I think 1.7.7.0 might be affected. However any 1.7.6.x or below should be fine.

Now that I think about the version numbers. 1.7.7.0 and 1.7.7.1 were released in rather quick succession. 
So, if I remember correctly, the day I wanted to upgrade, 7.1 was just released. Thus the upgrade went directly from 6.9 to 7.1. Maybe that is part of the problem, let me check some backups, to verify if this happens, when upgrading 6.9 -> 7.0 -> 7.1.

Let's hope this is just a very special corner case.

**Comment by hibatallahAouadni** [comment]
Hello @FredBrockstedt 

Thanks for your feedback!
Your issue is similar to these issues #22386 & #22611
Please check and feedback.

Thanks!

**Comment by FredBrockstedt** [comment]
Howdy @hibatallahAouadni,

thanks for keeping all the past issues in your memory.
#22386 is exactly the issue, with the slight deviation, that not all data our was affected, making this much harder to catch.
#22215 is also related and I think they came up with a very similar solution: https://github.com/PrestaShop/PrestaShop/issues/22215#issuecomment-748894921
I guess it would be smart to merge it to one of the previous issues.

#22611 doesn't seem to be related.

**Comment by hibatallahAouadni** [comment]
Hello @FredBrockstedt 

Thanks for your feedback and your PR :rocket: 
So, to sumarize the issue:

1. Install 1.7.6.9
2. Create a Customer Address without phone, company, address2, phone_mobile
3. Create an Order with this Customer
4. Upgrade your shop to 1.7.7.1
5. Try to open the Order => See error

Is that so?
Please check and feedback.

Thanks!

**Comment by FredBrockstedt** [comment]
hi @hibatallahAouadni 

I'm not entirely sure, it's actually necessary to install 1.7.6.9.
If the fields phone or address2 are optional, they are set to be nullable in the database.
Therefore someone creating an account without a mobile phone number for example, would cause a null in the database, which blows up the backend.
Since the address can be modified by user and backoffice, there are a lot more options for things to go wrong.
Altering the column to be not null, seems like a solid solution. So far nothing went wrong, but I'm not sure something fails, if it can't write a null value to a column if a field is optional?

**Comment by hibatallahAouadni** [comment]
Hello @FredBrockstedt 

> Install 1.7.6.9
Create a Customer Address without phone, company, address2, phone_mobile
Create an Order with this Customer
Upgrade your shop to 1.7.7.1
Try to open the Order => See error

When I follow these steps, I didn't manage to reproduce the issue :confused: 
Is there any detail you forget to mention?
Please check and feedback.

Thanks!

**Comment by hibatallahAouadni** [comment]
Hello @FredBrockstedt ,

Since we had no news from you for more than 20 days, I'll close this ticket to avoid cluttering up the backlog.
Please note that you can always create a new one if further information pops up.

Thanks!

**Comment by abcreationIT** [comment]
It seems not to be solved.
Still present on PrestaShop 1.7.7.4

I can't access the order page

**Comment by Davidorul** [comment]
Oh. Man. I wonder how many of these tickets exist ?

Soooo many problems with NULL values after upgrades... And all tickets get closed. What a shame.

**Comment by hibatallahAouadni** [comment]
Hello @Davidorul 

Could you please provide steps to reproduce the issue, because with the provided infos I didn't manage to reproduce the issue.
Then I will reopen the issue :wink: 

Thanks in advance :pray:

**Comment by Davidorul** [comment]
Thank you very much for your offer.

But as many others before me it was easier to go through the database and search / replace the NULL values than trying to setup a test-shop and re-do the upgrade to make this (maybe?) reproducable.

It pains me a bit as I'm sure countless others will run into the same issue. But I'm not a developer and I also don't have too much time for this. - We just needed to get our shop running again.

Just in case someone will also find this ticket and wonders.  - For us the final error was resolved after running:

`UPDATE `ps_product_attribute` SET `reference`= 0 WHERE `reference` IS NULL;`

(We had product combinations and variations that were causing the error.)
